File: flow_container/flow/flow/templates/js/validation.js
function validateUsername(username, errorField) {
    if (username.value === '' || username.value == null)
        errorField.textContent = "Username is required!";
    else if (username.value.length < 4)
        errorField.textContent = "Username must be at least 4 characters long!";
    else if (username.value.length > 15)
        errorField.textContent = "Username must be at most 15 characters long!";
    else if (!/^[a-zA-Z0-9]+$/.test(username.value))
        errorField.textContent = "Username must contain only alphanumeric characters!";
    else
        return true;
    username.classList.add('error');
    errorField.classList.add('show');
    return false;
}

function validateEmail(email, errorField) {
    if (email.value === '' || email.value == null)
        errorField.textContent = "Email is required!";
    else if (!/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/.test(email.value))
        errorField.textContent = "Email is invalid!";
    else
        return true;
    email.classList.add('error');
    errorField.classList.add('show');
    return false;
}

function validatePassword(password, confirmPassword, errorField) {
    if (password.value === '' || password.value == null)
        errorField.textContent = "Password is required!";
    else if (password.value.length < 5)
        errorField.textContent = "Password must be at least 5 characters long!";
    else if (password.value.length > 20)
        errorField.textContent = "Password must be at most 20 characters long!";
    else if (!/[A-Z]/.test(password.value))
        errorField.textContent = "Password must contain at least one upper-case letter!";
    else if (!/[a-z]/.test(password.value))
        errorField.textContent = "Password must contain at least one lower-case letter!";
    else if (!/[0-9]/.test(password.value))
        errorField.textContent = "Password must contain at least one digit character!";
    else if (password.value !== confirmPassword.value) {
        errorField.textContent = "Passwords do not match!";
        confirmPassword.classList.add('error');
    }
    else
        return true;
    password.classList.add('error');
    errorField.classList.add('show');
    return false;
}

function validateChatMessage(message, errorField) {
    if (message.value === '' || message.value == null)
        return false;
    else if (/^\s*$/.test(message.value)) {
        message.value = '';
        return false;
    }
    else if (message.value.length > 500)
        errorField.textContent = "Message must be at most 500 characters long!";
    else if (message.value.startsWith('GAME_INVITE='))
        message.value = message.value.substring(12);
    else
        return true;
    errorField.classList.add('show');
    return false;
}

File: flow_container/flow/api/validation.py
from django.contrib.auth.models import User
import re

def validate_alias(alias):
	if not alias:
		return "Alias is required!"
	elif len(alias) < 4:
		return "Alias must be at least 4 characters long!"
	elif len(alias) > 15:
		return "Alias must be at most 15 characters long!"
	elif not re.match("^[a-zA-Z0-9]+$", alias):
		return "Alias must contain only alphanumeric characters!"
	return None

def validate_username(username):
    if not username:
        return "Username is required!"
    elif len(username) < 4:
        return "Username must be at least 4 characters long!"
    elif len(username) > 15:
        return "Username must be at most 15 characters long!"
    elif not re.match("^[a-zA-Z0-9]+$", username):
        return "Username must contain only alphanumeric characters!"
    elif User.objects.filter(username=username).exists():
        return "Username already exists"
    return None

def validate_email(email):
    if not email:
        return "Email is required!"
    elif not re.match("^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$", email):
        return "Email is invalid!"
    return None

def validate_password(password, confirmPassword):
    if not password:
        return "Password is required!"
    elif len(password) < 5:
        return "Password must be at least 5 characters long!"
    elif len(password) > 20:
        return "Password must be at most 20 characters long!"
    elif not re.search("[A-Z]", password):
        return "Password must contain at least one upper-case letter!"
    elif not re.search("[a-z]", password):
        return "Password must contain at least one lower-case letter!"
    elif not re.search("[0-9]", password):
        return "Password must contain at least one digit character!"
    elif password != confirmPassword:
        return "Passwords do not match!"
    return None

def validate(data):
    if 'username' in data:
        error = validate_username(data['username'])
        if error:
            return error
    if 'email' in data:
        error = validate_email(data['email'])
        if error:
            return error
    if 'password' in data and 'confirm_password' in data:
        error = validate_password(data['password'], data['confirm_password'])
        if error:
            return error
    if 'password' in data and 'confirm_password' not in data:
        return "Confirm Password is required!"
    return None
